### PR TITLE
fix: (1) Add `StatcanDialogueDatasetRetrieval` (2) Fix `DRESModel.encode_conversations` to allow list of dictionaries

### PR DESCRIPTION
**Pull request by xhluca**
This pull request introduces two changes:
1. New dataset: `StatcanDialogueDatasetRetrieval`
2. Fix: Change `DRESModel.encode_conversations` to allow conversations composed of list of dictionaries, alongside lists of strings (e.g. Topicoqa). Also fix `batch_size` parameter in `DRESModel.encode_conversations` (b5141e8db5c51b32a16d4922058b5c25a1479338)

## Checklist for adding MMTEB dataset

Reason for dataset addition:

- [x] I have tested that the dataset runs with the `mteb` package.
- [x] I have run the following models on the task (adding the results to the pr). These can be run using the `mteb -m {model_name} -t {task_name}` command.
  - [x] `sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2`
  - [x] `intfloat/multilingual-e5-small`
- [x] I have checked that the performance is neither trivial (both models gain close to perfect scores) nor random (both models gain close to random scores).
- [x] If the dataset is too big (e.g. >2048 examples), considering using `self.stratified_subsampling() under dataset_transform()`
- [x] I have filled out the metadata object in the dataset file (find documentation on it [here](https://github.com/embeddings-benchmark/mteb/blob/main/docs/adding_a_dataset.md#2-creating-the-metadata-object)).
- [x] Run tests locally to make sure nothing is broken using `make test`. 
- [x] Run the formatter to format the code using `make lint`. 
- [x] I have added points for my submission to the [points folder](https://github.com/embeddings-benchmark/mteb/blob/main/docs/mmteb/points.md) using the PR number as the filename (e.g. `438.jsonl`).

## Other

- [x] change to s2p: https://github.com/embeddings-benchmark/mteb/pull/779#discussion_r1608761789
- [x] pin to specific version of dataset: https://github.com/embeddings-benchmark/mteb/pull/779#discussion_r1608768955
- [x] Test: `evaluation = MTEB(task_langs=["en", "de"]) # Only select datasets which are "en", "de" or "en-de"`

## Notes

> I have tested that the dataset runs with the `mteb` package.

To test it, I have used the following code:
```python
from mteb import MTEB
from mteb.evaluation.evaluators import DRESModel


class DummyModel:
    def encode(self, texts, *args, **kwargs):
        return texts


evaluation = MTEB(tasks=["StatcanDialogueDatasetRetrieval"])
evaluation.load_tasks_data()
sdd = evaluation.tasks[0]

q = sdd.queries["french"]["test"]["Q12210"]

doc_ids = list(sdd.relevant_docs["french"]["test"]["Q12210"].keys())
answer_1 = sdd.corpus["french"]["test"][doc_ids[0]]
answer_2 = sdd.corpus["french"]["test"][doc_ids[1]]


dres = DRESModel(model=DummyModel())

assert isinstance(q, list) is True, "Query is not a list"

encoded = dres.encode_conversations([q], batch_size=1)
print(encoded[0])
# user: bonjour; operator: Bonjour, je m'appelle Kelly C. Comment puis-je vous aider?; user: je cherches des données sur le secteur du porc canadien; user: je suis vraiment perdu dans le site; operator: Un instant; operator: Veuillez consulter l'hyperlien suivant: Production animale (filtre: porc) (https://www150.statcan.gc.ca/n1/fr/sujets/agriculture_et_alimentati; user: avez vous des données sur les exportations?; operator: un instant; user: en fait je veux mesurer l'importance de ce secteur en termes de production, de revenus, d'exportation et d'emploi.; operator: Activité commerciale internationale pour le code 1 du SH (https://www5.statcan.gc.ca/cimt-cicm/commodities-marchandises?lang=fra&ch +Live+animals+and+animal+products.&refMonth=7&refYr=2020&freq=6&countryId=999&usaState=0&provId=1&dataTransformation=0&searchStr=&monthStr=July); operator: et Activité commerciale internationale pour le code 2 du SH (https://www5.statcan.gc.ca/cimt-cicm/commodities-marchandises?lang=fra& +Live+animals+and+animal+products.&refMonth=7&refYr=2020&freq=6&countryId=999&usaState=0&provId=1&dataTransformation=0&searchStr=&monthStr=July); operator: Pour la production, je vous invite à consulter le premier hyperlien que j'ai partagé. Il y aurait des tableaux comme 32-10-0126-01 (https://w pour les exports, veuillez consulter les hyperliens ci-dessus; operator: pour les revenus et emplois, un instant; user: pour les exports, je vois des données mensuelles; user: est il possible d'avoir des données annuelles, un cumul?; operator: Oui, vous avez l'option de changer la fréquence; operator: Est-ce que vous voyez l'option?; user: exemple pour 2019, je ferais quoi pour avoir les données de toute l'année; operator: Vous devez changer l'année à 2019, et la fréquence à annuel. Vous devez ensuite extraire les données. Exemple: Tableau 980-0002 (https://w lang=fra&getSectionId()=1&dataTransformation=0&refYr=2019&refMonth=7&freq=12&countryId=0&getUsaState()=0&provId=1&retrieve=Extraire&country=null&trad; user: et le mois, je vais choisir quoi; operator: pour le revenu, veuillez consulter le tableau suivant: 32-10-0136-01 (https://www150.statcan.gc.ca/t1/tbl1/fr/tv.action?pid=3210013601&re; operator: Vous pouvez selectionner le mois que vous voulez.; user: super merci; user: pour l'emploi; user: etes vous avec moi?; operator: oui, un instant; user: si je veux connaitre la place du canada dans le monde en matière de porc, en termes de production et d'exportation; operator: Je suis toujours à la recherche de vos données. Merci de votre patience.; user: un gros merci
```

> I have checked that the performance is neither trivial (both models gain close to perfect scores) nor random (both models gain close to random scores).

Looking at the scores (recall at k):
```
"recall_at_1": 0.03704,
"recall_at_3": 0.07407,
"recall_at_5": 0.09877,
"recall_at_10": 0.15895,
"recall_at_20": 0.17747,
"recall_at_100": 0.39738,
"recall_at_1000": 0.72222,
```

we can  see that the recall at k from 1 to 1000 consistently, which means it is a fairly challenging task (r@10 is only 0.15) but non-random.

### CODE REVIEW

**Comment by xhluca** [comment]
@orionw @vaibhavad would love if you can review this PR! There's a few changes i want to add but feel free to leave feedback now if you wish!

**Comment by orionw** [comment]
The current code looks like a great start -- the remainder todos are just filling in the things marked TBD and adding results/points.

File: mteb/tasks/Retrieval/multilingual/StatcanDialogueDatasetRetrieval.py
Comment (vaibhavad) [inline mteb/tasks/Retrieval/multilingual/StatcanDialogueDatasetRetrieval.py:76]: Does StatsCan have multiple matching gold documents/table or single?

File: mteb/tasks/Retrieval/multilingual/StatcanDialogueDatasetRetrieval.py
Comment (vaibhavad) [inline mteb/tasks/Retrieval/multilingual/StatcanDialogueDatasetRetrieval.py:70]: Might be better to pin down to a specific version

File: mteb/tasks/Retrieval/multilingual/StatcanDialogueDatasetRetrieval.py
Comment (vaibhavad) [inline mteb/tasks/Retrieval/multilingual/StatcanDialogueDatasetRetrieval.py:47]: Current for conversational QA the expectation is that query is list of strings, however in StatsCan each query is a dictionary with "role" and "content" keys. This format is definitely better as it allows for dynamic number of utterances per turn and also supports multi-user-agent conversations. However some changes will be required to make it work.

1. The [check](https://github.com/embeddings-benchmark/mteb/blob/eea434adcce71cabf9ebfbecd111f929eba68434/mteb/evaluation/evaluators/RetrievalEvaluator.py#L79) that decides between `encode_queries` and `encode_conversations` will change.
2. The [default method for converting queries to input string](https://github.com/embeddings-benchmark/mteb/blob/eea434adcce71cabf9ebfbecd111f929eba68434/mteb/evaluation/evaluators/RetrievalEvaluator.py#L402) will also change.

**Review by vaibhavad — CHANGES_REQUESTED** [review]

File: mteb/tasks/Retrieval/multilingual/StatcanDialogueDatasetRetrieval.py
Comment (vaibhavad) [reply mteb/tasks/Retrieval/multilingual/StatcanDialogueDatasetRetrieval.py:47]: I can also change TopiOCQA loading accordingly once the above mentioned changes are done.

File: mteb/tasks/Retrieval/multilingual/StatcanDialogueDatasetRetrieval.py
Comment (xhluca) [reply mteb/tasks/Retrieval/multilingual/StatcanDialogueDatasetRetrieval.py:73]: TODO

File: mteb/tasks/Retrieval/multilingual/StatcanDialogueDatasetRetrieval.py
Comment (xhluca) [reply mteb/tasks/Retrieval/multilingual/StatcanDialogueDatasetRetrieval.py:70]: TODo

File: mteb/tasks/Retrieval/multilingual/StatcanDialogueDatasetRetrieval.py
Comment (xhluca) [reply mteb/tasks/Retrieval/multilingual/StatcanDialogueDatasetRetrieval.py:47]: Maybe this would be better as new issue?

File: mteb/tasks/Retrieval/multilingual/StatcanDialogueDatasetRetrieval.py
Comment (xhluca) [reply mteb/tasks/Retrieval/multilingual/StatcanDialogueDatasetRetrieval.py:112]: Oh I just carried it from my previous one. Should I just use _LANGS instead?

File: mteb/tasks/Retrieval/multilingual/StatcanDialogueDatasetRetrieval.py
Comment (xhluca) [reply mteb/tasks/Retrieval/multilingual/StatcanDialogueDatasetRetrieval.py:76]: Single document per example

**Comment by xhluca** [comment]
Seems like language filtering does not work, or i'm running the wrong command:

```python
from mteb import MTEB
evaluation = MTEB(task_langs=["de"])
print(evaluation.available_tasks)  # StatcanDialogueDatasetRetrieval will appear
```

**Comment by vaibhavad** [comment]
@xhluca,

I don't think this is the current API. It is show all tasks, regardless of category, language. Can you try `evaluation.print_selected_tasks()` instead?

**Comment by xhluca** [comment]
Thanks, seems print_selected_tasks works.

File: mteb/tasks/Retrieval/multilingual/StatcanDialogueDatasetRetrieval.py
Comment (vaibhavad) [reply mteb/tasks/Retrieval/multilingual/StatcanDialogueDatasetRetrieval.py:47]: Changes to be made [here](https://github.com/embeddings-benchmark/mteb/blob/eea434adcce71cabf9ebfbecd111f929eba68434/mteb/evaluation/evaluators/RetrievalEvaluator.py#L402)

File: mteb/tasks/Retrieval/multilingual/StatcanDialogueDatasetRetrieval.py
Comment (xhluca) [reply mteb/tasks/Retrieval/multilingual/StatcanDialogueDatasetRetrieval.py:78]: Sorry, there might be more than 1 in some cases.

**Comment by xhluca** [comment]
@vaibhavad @orionw As discussed in [this comment](https://github.com/embeddings-benchmark/mteb/pull/779#discussion_r1608812884), a fixed was needed for `DRESModel.encode_conversations` (i.e. `convert_conv_history_to_query`) in order to allow a conversation to be represented as a list of dict rather than a list of string. This format is directly compatible with huggingface-style conversations (see this tutorial). See this commit for the details: https://github.com/embeddings-benchmark/mteb/pull/779/commits/b5141e8db5c51b32a16d4922058b5c25a1479338

Let me know if this fix makes sense (and whether I should tag this PR as bug fix or something else).

File: mteb/tasks/Retrieval/multilingual/StatcanDialogueDatasetRetrieval.py
Comment (orionw) [inline mteb/tasks/Retrieval/multilingual/StatcanDialogueDatasetRetrieval.py:72]: This should be the git commit hash of the dataset, if possible.

File: mteb/evaluation/evaluators/RetrievalEvaluator.py
Comment (orionw) [inline mteb/evaluation/evaluators/RetrievalEvaluator.py:404]: This changes looks good to me.  Can you update the Tapioca dataset as well to handle this / ensure it still runs?

**Review by orionw — CHANGES_REQUESTED** [review]
Have some minor comments, but I like the changes.  

No need to tag the PR (not sure we use that?) but definitely add points for the dataset, 1-2 for a bugfix, and also the reviewers.

**Comment by xhluca** [comment]
Huh, seems tests failed after i merged changes from `main` to this branch...

File: mteb/evaluation/evaluators/RetrievalEvaluator.py
Comment (xhluca) [reply mteb/evaluation/evaluators/RetrievalEvaluator.py:404]: @orionw good catch - seems i introduced a subtle bug when modifying this function. it has been fixed and topiocqa should work now.



Here's the code:

```python

from mteb import MTEB
from mteb.evaluation.evaluators import DRESModel


class DummyModel:
    def encode(self, texts, *args, **kwargs):
        return texts


evaluation = MTEB(tasks=["TopiOCQA"])
evaluation.load_tasks_data()
sdd = evaluation.tasks[0]

q = sdd.queries["validation"]["205-11"]

dres = DRESModel(model=DummyModel())

assert isinstance(q, list) is True, "Query is not a list"

encoded = dres.encode_conversations([q], batch_size=1)
print(encoded[0])
```

Which should return:

```
what was one of the provisions of the petition of right (1628); No person should be forced to provide a gift, loan or tax without an Act of Parliament.; what does the term "petition of right" refer to?; It is an English constitutional document setting out specific individual protections against the state.; who formulated it?; Charles I; what kinds of provisions does it have?; No person should be forced to provide a gift, loan or tax without an Act of Parliament. No free individual should be imprisoned or detained unless a cause has been shown. Soldiers or members of the Royal Navy should not be billeted in private houses without the free consent of the owner.; which period marked the age of sail for the navy mentioned above?; During the reign of Henry VIII.; do they have an escort fleet?; Yes, there are six Type 45 destroyers and 13 Type 23 frigates in active service.; what is type 45?; It is also known as the D or "Daring" class. It is a class of six guided missile destroyers built for the United Kingdom's Royal Navy.; did its engine face any trouble?; On occasion there have been near-complete power generation failures, temporarily disabling propulsion, power generation for weapons, navigational systems and other purposes, leaving the ships vulnerable to "total electric failure".; how strong was their air defense?; The Royal Navy describes the destroyers' mission as being "to shield the Fleet from air attack". It is equipped with the Sea Viper air-defence system utilizing the SAMPSON active electronically scanned array multi-function radar and the S1850M long-range radar. PAAMS is able to track over 2,000 targets and simultaneously control and coordinate multiple missiles in the air at once, allowing a large number of tracks to be intercepted and destroyed at any given time.; how does the first mentioned radar function?; The SAMPSON multi function radar can detect all types of targets out to a distance of 400 km, and is capable of tracking hundreds of targets at any one time.; who developed it?
```

**Comment by orionw** [comment]
> Huh, seems tests failed after i merged changes from main to this branch...

I think those are fixed by #803 if you merge in main again, agree they weren't caused by this PR

**Review by orionw — APPROVED** [review]
Once the tests pass, will enable automerge

**Comment by xhluca** [comment]
@vaibhavad can you approve if everything has been covered?

**Review by vaibhavad — APPROVED** [review]

**Comment by vaibhavad** [comment]
Approved and merged, thanks for the great work @xhluca!

**Pull request merged**
